DECOMPRESSION FPGA DESIGN (GZIP and SNAPPY)

tyoungsc opened the pull request:
Internal Review

tyoungsc committed:
First commit for gunzip

tyoungsc committed:
First version of code working in HW

tyoungsc committed:
First version without profiling that functions in HW

tyoungsc committed:
Code working with ac_ints

tyoungsc committed:
Fixed bug in parsing of first code table
Improved performance of huffman decoder via shannonization and manually increasing II on non-critical loop kernels to avoid Fmax bottlenecks
First version working in HW with small to large files

tyoungsc committed:
Optimization:
  Restructred if-statement checking symbol in main processing loop to check literal and distance symbol in parallel based on state. This removes a 9-bit select from critical path and uses less area for the distance compares and computations since it can use 5-bits rather than 9-bit.

Validated in HW.

tyoungsc committed:
Removed unused state from main processing loop enum
Added file I forgot to add before

tyoungsc committed:
Optimization:
  Use ac_int to build values read from ByteBitStream

tyoungsc committed:
Optimization:
  Store "space" in ByteBitStream to avoid subtraction
  Used unsigned char in ByteBitStream to reduce bits for tracking size and space

tyoungsc committed:
Comments on main processing loop

tyoungsc committed:
Optimization:
  Decode literal length and and dynamic extra bits in a single iteration

tyoungsc committed:
Optimization:
  Greatly improved LZ77 decoder by reducing to a single load/store site for history memory
  Used caching method to break read/write dep and improve II
  Shannonize a few variables to improve critical path

tyoungsc committed:
Comments

tyoungsc committed:
Read exact output file size in host from file to size output buffer
Added method to mp_math for rounding up to nearest power

tyoungsc committed:
Optimization:
  Decoding distance and extra distance bits in single iteration. Can now decoded length distance pair in 2 iterations
  Cleaned up code now that we only need 2 iterations to parse length distance pair

tyoungsc committed:
Optimization:
  Disconnected reading in new data and computing on bits in main loop

tyoungsc committed:
Optimizations:
  HuffmanDecoder kernel notify downstream kernel earlier
  Reduced width of counters and compares

tyoungsc committed:
Cleaned up readuint functions to generalize and template them
Aliased ac_uint<num> = ac_int<num, false> because we only use unsigned int
Other code cleanup and comments
Optimizations:
  Parsed first table using same method as second table (storing first and last codes + base indices)
  Improved parsing of first table so it could be pipelined
  Removed initialization of code tables that was unnecessary
  Chose smaller widths for many variables to fit exactly what is needed

tyoungsc committed:
Code cleanup
Optimizations:
  Reduced width of variables from first table
  Tighter bound on maximum number of dynamic bits from ByteBitStream

tyoungsc committed:
Switched from a ring buffer to a shift register for ByteBitStream

tyoungsc committed:
Fixed bug with {0} initialization of ac_int array (will file bug)
Changed main file to default to 2 runs for emulation

tyoungsc committed:
Updating bit shift register length

tyoungsc committed:
Same as last commit ...

tyoungsc committed:
Updated CMake file to include option to ignore default seed
Cleaned up ByteBitStream a bit. Used template args to pick size, max dynamic read bits, and max shift bits
Changed to using a dynamic and static version of ReadUInt to make things more readable
Removed Unused functions

Optimization:
  Use maximum dynamic read and shift bits, set the sizes of args to ReadUInt and Shift
  Split determination of shortest match length into two for-loops, in parallel and reduce from there

tyoungsc committed:
Reverted lots of the previous commit because it didn't help
Optimization:
  Used slc method for ac_int in ByteBitStream
  In decompressor main loop, find shortest match length first, use it as an index to find the base index and offset

tyoungsc committed:
Trying to use if-else_of rather than two ifs in main loop

tyoungsc committed:
Added method to count trailing 0s
Used signed integers for unrolled loops
Added simulator target
Went back to if-else_if (not good perf)

tyoungsc committed:
Set II of non-critical loop

tyoungsc committed:
Actually split if_elseif in main loop
Renamed function ctz to get trailing zeros

tyoungsc committed:
Fixed bit width calculations in ByteBitStream

tyoungsc committed:
Implemented special version of ctz for 15 bits
Fixed main processing loop to II=4 (for testing)

tyoungsc committed:
Got rid of tree for ctz, didn't help Fmax (hurt slightly)
Used set_slc for new byte of ByteBitStream
More accurate with ac_int and constants in Huffman decoder

tyoungsc committed:
For add/subtracts with ac_int and constant, wrap constant with ac_int<bits> to avoid getting 33-bit integer add/sub
Removed divide and mods from critical path

tyoungsc committed:
Forcing II=2

tyoungsc committed:
Removed fixed II on main loop
Removed a 1-bit AND from the critical path (not necessary to guard non-blocking pipe read)

tyoungsc committed:
Allowed LZ77 to do multiple elements per cycle
Added LiteralPacker kernel to take multiple elements per cycle from LZ77 and align them
Removed USM debugger class; shouldn't need it anymore
Created function to launch all decompress kernels (called in main())
Made design fail if inflated_count != expected size

tyoungsc committed:
First version of multi element per cycle LZ77 working in emulation

tyoungsc committed:
Changed multi-element LZ77 to be better

tyoungsc committed:
Improved LZ77 multi-element per cycle
Verified in sim

tyoungsc committed:
Added shift register caches for each history buffer
Verified in simulation

tyoungsc committed:
Use NTuple and UnrolledLoop to workaround bugs in compiler with banking
Added tuple.hpp (for NTuple)
Matched simulator flags to hardware flags

tyoungsc committed:
Reduced cache depth for LZ77

tyoungsc committed:
Trying out new ctz method
Created constexpr Shift for ByteBitstream to shift a constant amount
Initialized ByteBitStream buffer with 0s

tyoungsc committed:
Reduced width of subtractions on critical path

tyoungsc committed:
Added first support for uncompressed and statically compressed blocks

tyoungsc committed:
Undid special case ctz. Didn't help.
Got rid of templated Shift. Didn't help.
Added method to align byte stream to byte boundary (for uncompressed blocks)
Cleaned up dead code.

tyoungsc committed:
Huge code cleanup. Snake case for files, separated kernels into different files, comments, ...
Created special LZ77 for 1 element per cycle
Created a simple CRC32 function to validate output data

tyoungsc committed:
Trying new method using ctz and ctz+1 in parallel
code cleanup

tyoungsc committed:
Slight improvement to CTZ

tyoungsc committed:
Renamed literal_stacker to byte_stacker and changes accordingly

tyoungsc committed:
fixed mod_lut calculation

tyoungsc committed:
Removed constraint on number of runs from command line

tyoungsc committed:
Fixed output message english

tyoungsc committed:
Code cleanup

tyoungsc committed:
Code formatting

tyoungsc committed:
Found bugs in parsing uncompressed blocks. Fixed them.
Code cleanup.

tyoungsc committed:
Updating samples.json and VS files

tyoungsc committed:
Stripped down README to start working on it

tyoungsc committed:
Code cleanup.
Renamed header reader to metadata reader

tyoungsc committed:
Added method to set literals per cycle from cmake

tyoungsc committed:
Changing default runs for HW

tyoungsc committed:
2 literals per cycle in main loop
code cleanup

tyoungsc committed:
Try forcing II=2

tyoungsc committed:
Removed parsing two literals per cycle

tyoungsc committed:
Rebased with master and used files from new "include" directory (deleted them from this project)

tyoungsc committed:
Code cleanup

tyoungsc committed:
First version of README done
Small change to ByteBitStream to do more efficient check
Comments

tyoungsc committed:
Change main file to run more tests by default but also allow user to just decompress a file
Added test data

tyoungsc committed:
Changed usage print

tyoungsc committed:
Fixed bug in static huffman decompression

tyoungsc committed:
Changed printed output
Updated README
Generated a GUID for samples.json and added description

tyoungsc committed:
Removed ByteStacker when literals_per_cycle is 1
Changed some variable names to be more clear

tyoungsc committed:
Setting default literals_per_cycle to 4

tyoungsc committed:
Updating output and README
Updated picture for Huffman decoder

tyoungsc committed:
README updated

tyoungsc commented:
@mdbtucker

yuguen commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,420 @@
+  // the mod_lut_init ROM that was computed above
+  [[intel::fpga_register]]
+  HistBufBufIdxT mod_lut[literals_per_cycle - 1][literals_per_cycle - 1];
+  for (int y = 0; y < (literals_per_cycle - 1); y++) {
```
Would loop coalescing make a difference here?
I don't understand why in your report it says that this is limiting the fmax

tyoungsc replied:
I may change this code soon based on what Mike showed in our meeting last week.

yuguen commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,420 @@
+      HistBufBufIdxT shuffle_vec[literals_per_cycle];
+      
+      #pragma unroll
+      for (int i = 0; i < literals_per_cycle; i++) {
```
Should we stick to "UnrolledLoop" or "pragma unroll" rather than having a mix of both?

tyoungsc replied:
I agree. I went originally with unrolled for, but needed UnrolledLoop in this case to work around these bugs:
https://hsdes.intel.com/appstore/article/#/14015867379
https://hsdes.intel.com/appstore/article/#/14015870207

mdbtucker replied:
I think this comment has been resolved?

yuguen commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,420 @@
+      // if we get a length distance pair we will read 'pipe_data.data.length'
+      // bytes starting at and offset of 'dist'
+      history_counter = pipe_data.data.length;
+      reading_history = pipe_data.data.is_copy && data_valid;
```
You usually use & instead of && - is this on purpose?

tyoungsc replied:
I am using && here, no? These are booleans so using an '&' would produce a warning.

yuguen replied:
Right but Paul was mentioning at some point that && has a sequential evaluation which was unwanted for HW generation (I actually never tried using & with booleans)

tyoungsc replied:
Spoke offline and the '&&' avoids a compiler warning and does not affect performance here.

yuguen commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/common.hpp:
```diff
@@ -0,0 +1,141 @@
+auto CTZ(const ac_int<bits, is_signed>& in) {
+  static_assert(bits > 0);
+  constexpr int out_bits = fpga_tools::Log2(bits) + 1;
+  //ac_uint<out_bits> ret(bits);
```
Remove comment?

yuguen commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/common.hpp:
```diff
@@ -0,0 +1,141 @@
+//
+template<int bits, bool is_signed>
+auto CTZ(const ac_int<bits, is_signed>& in) {
+  static_assert(bits > 0);
```
Looking at the report, the loop that gets the lowest fmax is in the Huffman decoder and the feedback path uses this CTZ. The feedback path in the throughput analysis reports that it needs one cycle per bit select... I'll reach out to you about this

tyoungsc replied:
Sounds good! You may remember me asking about this in the past. I tried a lot of things to make it better (using a tree and other interesting techniques), but with no luck

tyoungsc replied:
Spoke offline. Nothing to do here.

tyoungsc committed:
Made CTZ behave correctly when passed 0 as argument
Reduced code for 'mod_lut' to just use constexpr array in registers
Removed complicated >= 8 code that the compiler does for us in byte_bit_stream

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+# GZIP Decompression
+This DPC++ reference design demonstrates an GZIP decompression design on an FPGA.
```
"This gzip decompression design does gzip decompression" :-)  How about this:
```suggestion
This DPC++ reference design implements GZIP decompression on an FPGA.  This includes LZ77 decoding and Huffman decoding with static or dynamic huffman tables. 
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+| OS                                | Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
+| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel Xeon&reg; CPU E5-1650 v2 @ 3.50GHz (host machine)
+| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | Creating an efficient GZIP decompression engine on an FPGA.
```
```suggestion
| What you will learn               | How to implement an efficient GZIP decompression engine on an FPGA.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+| Time to complete                  | 1 hour
+
+## Purpose
+This FPGA reference design demonstrates an efficient GZIP decompression engine. See the [Additional Design Information Section](#additional-design-information) for more information on GZIP and DEFLATE de/compression for the FPGA.
```
```suggestion
This FPGA reference design demonstrates an efficient GZIP decompression engine. See the [Additional Design Information Section](#additional-design-information) for more information on GZIP and DEFLATE compression and decompression.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+# GZIP Decompression
```
General comment on this file: there was a recent PR for all the FPGA Readmes: https://github.com/oneapi-src/oneAPI-samples/pull/858  Can you make sure these changes are incorporated into this README?

tyoungsc replied:
I checked out the recent change. Does all of that information really belong in every README? It is generic to all programs, could it not go in a higher level README?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+All kernels have finished for run 4
+Execution time: 7.12956 ms
+Output Throughput: 561.045 MB/s
+Compression Ratio: 1021.26
```
Maybe this should be expressed explicitly as a ratio (xxx:1)?

Also, is this number right?  That file is compressed 1021:1 ???

tyoungsc replied:
I can switch to a percentage, or use a ratio. What do you think?
and yes. It is a max throughput test.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
````diff
@@ -0,0 +1,281 @@
+```
+
+NOTE: When running on the FPGA emulator, the *Execution time* and *Throughput* do not reflect the design's actual hardware performance. <br>
+NOTE: The throughput of the decompression engine is highly dependant on the compression ratio of the input file, more specifically on the amount of LZ77 compression. The throughput test reports the *maximum* throughput, since it is decompressing a highly compressed file. When decompressing other files, it is expected that the throughput will be lower.
````
```suggestion
NOTE: The throughput of the decompression engine is highly dependant on the compression ratio of the input file, more specifically on the amount of LZ77 compression. The file chosen for this test has a very high amount of LZ77 compression, and was chosen to demonstrate the *maximum* throughput achievable with this implementation. When decompressing other files, the throughput may be somewhat lower.
```
Can we put a lower bound on how much lower?  'In the worst case, as much as xx% (50%?) lower'

tyoungsc replied:
The lowest would be a completely uncompressed file, which would be like 90% decrease.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+
+| File                            | Description 
+|:---                             |:---
+|`main.cpp`                       | Contains the `main()` function and the top-level launching, validation, and performance measurements.
```
```suggestion
|`main.cpp`                       | Contains the `main()` function which launches the kernels, validates the results, and measures performance.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+| File                            | Description 
+|:---                             |:---
+|`main.cpp`                       | Contains the `main()` function and the top-level launching, validation, and performance measurements.
+|`byte_bit_stream.hpp`            | A bitstream class that has 8-bits (a byte) written in and a dynamic number of bits read.
```
```suggestion
|`byte_bit_stream.hpp`            | A bitstream class that accepts one byte (8 bits) at a time and allows a variable number of bits to be read out on each transaction.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+|:---                             |:---
+|`main.cpp`                       | Contains the `main()` function and the top-level launching, validation, and performance measurements.
+|`byte_bit_stream.hpp`            | A bitstream class that has 8-bits (a byte) written in and a dynamic number of bits read.
+|`byte_stacker.hpp`               | A kernel that accepts between 0 and N elements per cycle, buffers N elements locally, and writes out N elements when possible.
```
```suggestion
|`byte_stacker.hpp`               | A kernel that accepts between 0 and N elements per cycle and combines them to output N elements at a time.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+|`main.cpp`                       | Contains the `main()` function and the top-level launching, validation, and performance measurements.
+|`byte_bit_stream.hpp`            | A bitstream class that has 8-bits (a byte) written in and a dynamic number of bits read.
+|`byte_stacker.hpp`               | A kernel that accepts between 0 and N elements per cycle, buffers N elements locally, and writes out N elements when possible.
+|`common.hpp`                     | Functions common across the design.
```
'...such as. . .'

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+|`simple_crc32.hpp`               | A simple implementation of CRC-32 calculation. This is used to validate the output of the GZIP engine.
+
+### GZIP and DEFLATE
+GZIP is specific implementation of the DEFLATE protocol. The structure of a GZIP file is illustrated in the figure below. It starts with a variable length, byte-aligned header of 10 or more bytes. Then follows the data payload, which is 1 or more DEFLATE compressed blocks. After the DEFLATE blocks, there may be some small amount of padding to realign to a byte boundary, and finally an 8-byte GZIP ...
```
```suggestion
GZIP is a specific implementation of the DEFLATE protocol. The structure of a GZIP file is illustrated in the figure below. It starts with a variable length, byte-aligned header of 10 or more bytes. Then follows the data payload, which is 1 or more DEFLATE compressed blocks. After the DEFLATE blocks, there may be some small amount of padding to realign to a byte boundary, and finally an 8-byte GZIP footer, which contains the CRC-32 and size (in bytes) of the uncompressed data. For more details on the GZIP file format, [see here](https://en.wikipedia.org/wiki/Gzip)
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+
+<img src="gzip_file_format.png" alt="gzip_file_format" width="400"/>
+
+The DEFLATE compression algorithm performs LZ77 Encoding followed by Huffman encoding (decompression decodes in the opposite order). For more information on the DEFLATE format and LZ77 and Huffman encoding/decoding, [start here](https://en.wikipedia.org/wiki/Deflate).
```
```suggestion
The DEFLATE compression algorithm performs LZ77 Encoding followed by Huffman encoding. Therefore, decompression decodes in the opposite order. For more information on the DEFLATE format and LZ77 and Huffman encoding/decoding, [start here](https://en.wikipedia.org/wiki/Deflate).
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+# GZIP Decompression
```
Please have someone in Tech Comms review this README

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+
+
+### GZIP Decompression FPGA Design
+The image below summarizes the full streaming GZIP decompression design. The orange kernels on the FPGA in the dashed box are the *actual* streaming GZIP decompression engine. The *Producer* and *Consumer* kernels are only used to stream data into and out of the GZIP decompression engine from device memory, respectively.
```
```suggestion
The image below summarizes the full streaming GZIP decompression design. The orange kernels on the FPGA in the dashed box make up the streaming GZIP decompression engine. The *Producer* and *Consumer* kernels are only used to stream data into and out of the GZIP decompression engine from and to device memory, respectively.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+
+<img src="gzip_decompression_engine_full.png" alt="gzip_decompression_engine_full" width="700"/>
+
+The figure below illustrates the GZIP decompression engine in isolation. It delineates the streaming GZIP decompressor from the streaming DEFLATE decompressor. The `GZIPMetaDataReader` kernel parses and strips away the GZIP header and footer data and forwards the consecutive DEFLATE blocks to the DEFLATE decompression engine.
```
I think I made this comment on your slides as well, but I'd just add the DEFLATE dotted line box to the diagram above and skip this second diagram.  Your call though.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+#### Huffman Decoder
+The Huffman decoder streams in DEFLATE compressed blocks, a byte at a time, and streams out either literals (8-bit characters) or {length, distance} pairs. The Huffman decoder is not byte-aligned; it decodes the input stream bits at a time. To do this efficiently, the decoder uses the `ByteBitStream` class, which streams in bytes at a time but gives the decoder access to all of the bits in parall ...
+
+The structure of the Huffman decoder is shown in the image below. The decoder examines the next 15 bits in the bit stream and checks for matches of length 1-15 in parallel to create a 15-bit *valid bitmap*, where a 1 at bit `i` indicates that the code of length `i` matches a code in the Huffman table. For example, a value of `0b000000000000101` would indicate that there is a match of length 1 and ...
```
```suggestion
The structure of the Huffman decoder is shown in the image below. The decoder examines the next 15 bits in the bit stream and checks for matches in the Huffman table of lengths 1-15 in parallel to create a 15-bit *valid bitmap*, where a 1 at bit `i` indicates that the code of length `i` matches a code in the Huffman table. For example, a value of `0b000000000000101` would indicate that there is a match of length 1 and 3.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+
+The structure of the Huffman decoder is shown in the image below. The decoder examines the next 15 bits in the bit stream and checks for matches of length 1-15 in parallel to create a 15-bit *valid bitmap*, where a 1 at bit `i` indicates that the code of length `i` matches a code in the Huffman table. For example, a value of `0b000000000000101` would indicate that there is a match of length 1 and ...
+
+To decode the next symbol, the decoder finds the *shortest* of the 15 matches that were computed in parallel. To do so, it performs a `CTZ` (**c**ount **t**railing **z**eros) operation on the valid bitmap and adds one to it, which is equivalent to finding the shortest match length. Once the shortest match length is found, the bit stream is shifted by that many bits, the symbol for that match is l ...
```
```suggestion
To decode the next symbol, the decoder finds the *shortest* of the 15 matches that were computed in parallel. To do so, it performs a `CTZ` (**c**ount **t**railing **z**eros) operation on the valid bitmap and adds one to it, which is equivalent to finding the shortest match length. Once the shortest match length is found, the bit stream is shifted by that many bits, the symbol for that match is determined from the Huffman table, and the process repeats. If the decoded symbol is a literal, then the literal is written to the output. If the decoded symbol is a length, then no output is generated in this iteration; the length is saved and the next iteration of the loop decodes the distance and generates the {length, distance} pair output.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+The Huffman decoder decodes a literal in 1 iteration of the main loop, and a {length, distance} pair in 2 iterations. When decoding a {length, distance} pair, the length code can be 1-15 bits and, based on the decoded length, 0-5 extra bits. Similarly, the distance code can be 1-15 bits and 0-15 extra bits. To decode a {length, distance} in 2 iterations, the Huffman decoder looks at 30 bits in pa ...
+
+#### LZ77 Decoder
+The input to the LZ77 decoder is a stream of commands that are either an 8-bit literal or a {length, distance} pair; the output of the Huffman decoder. The output is a stream of literals (8-bit characters). The LZ77 decoder keeps a 32K *history buffer*; a history of the last 32K literals it has streamed to the output. 
```
Is 'commands' the right term here?  Maybe 'symbols' is better?
```suggestion
The input to the LZ77 decoder is a stream of symbols(?) that are either an 8-bit literal or a {length, distance} pair. The output is a stream of literals (8-bit characters). The LZ77 decoder keeps a 32KB *history buffer* which stores the last 32K literals it has streamed to the output. 
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
````diff
@@ -0,0 +1,281 @@
+
+Let's assume that to start the LZ77 decoder received 14 consecutive literals and wrote them to the output and history buffers in cyclical order (i.e. to buffers `0, 1, 2, 3, 0, 1, 2, 3, ...`).
+```
+Characters:    E T H O M E P H O N E
````
Really?  ET HOME PHONE, not ET PHONE HOME?

tyoungsc replied:
https://www.youtube.com/watch?v=iyFijjikkeM

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md:
```diff
@@ -0,0 +1,281 @@
+
+In general, this process is repeated `ceil(length/N)` times. When `length` is not a multiple of `N`, the `valid_count` is used to tell the downstream kernel how many of the `N` literals are valid. There is an edge case when the `distance` in the {length, distance} pair is less than `N`. This is handled in the code, but is omitted in this document for brevity.
+
+The history buffers have a read-after-write dependency since, when reading from the history buffers, the LZ77 decoder is also writing back to them. The history buffers are implemented in on-chip RAMs and therefore have a read latency that results in an increased loop II. To fix this II issue, we use a shift-register cache to track in-flight writes to the history buffers. This technique is describ ...
```
Please include a link to the tutorial.

tyoungsc replied:
We haven't done in any other reference designs that cite other tutorials. E.g. https://github.com/oneapi-src/oneAPI-samples/blob/master/DirectProgramming/DPC%2B%2BFPGA/ReferenceDesigns/qrd/README.md

Should these all be changed?

mdbtucker replied:
Your call now that you own the samples.  Whatever you decide, I made a similar comment in @yuguen-intel 's review on his fork, so let him know either way

mdbtucker reviewed (commented):
Just went over the README so far, will review the code later

tyoungsc committed:
Addressing Mike's README comments

tyoungsc committed:
Clang format and update to some manually set IIs

tyoungsc committed:
Manually fused two loops in the Huffman decoder to reduce the overhead of parsing the tables

tyoungsc committed:
Templated lz77 decoder on maximum history distance

tyoungsc committed:
Small grammar

anton-intel committed:
Intel Modin Getting Started Sample. Readme edits. (#892)

- Used a new template.
- Edited the text.
- Added information on how to create a kernel to run the notebook in the DevCloud.

tyoungsc committed:
Fixing lambda spelling

yuguen committed:
FPGA QRD seed update (#894)

* update qrd seeds to meet 2022.2 target

Signed-off-by: Yohann Uguen <yohann.uguen@intel.com>

* Updating seeds for 2022.2

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/sample.json:
```diff
@@ -0,0 +1,55 @@
+      {
+        "id": "fpga_emu",
+        "steps": [
+          "mkdir build",
```
You haven't included the latest changes to the .json files to avoid long path problems.  I feel like there might have been other CI test changes too - please compare against an up-to-date .json file and make sure you have all the latest 'standard' changes to the CI tests.

tyoungsc replied:
Done. Also done for Snappy. Thanks.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_bit_stream.hpp:
```diff
@@ -0,0 +1,98 @@
+#include <CL/sycl/INTEL/ac_types/ac_int.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+#include "constexpr_math.hpp"
```
Comment location of this file in the common include directory

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_bit_stream.hpp:
```diff
@@ -0,0 +1,98 @@
+//
+// A stream of bits that is filled with a byte at a time
+//
+template <int bits, int max_dynamic_read_bits, int max_shift_bits>
```
I think a comment for each template parameter would be helpful.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_bit_stream.hpp:
```diff
@@ -0,0 +1,98 @@
+//
+// A stream of bits that is filled with a byte at a time
+//
+template <int bits, int max_dynamic_read_bits, int max_shift_bits>
```
I like the convention of starting non-type template parameters (ie ints) with k_, to indicate that within the code they can be treated as constants.  I'd really prefer kBits to k_bits, but that would violate Google naming conventions.  I'm not insisting on this, just a suggestion, feel free to ignore if you disagree.

tyoungsc replied:
I will change to k_ prefix.

tyoungsc replied:
I changed my mind :\. I am going to keep it as-is

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_bit_stream.hpp:
```diff
@@ -0,0 +1,98 @@
+  static_assert(max_shift_bits > 0);
+  static_assert(max_shift_bits <= bits);
+
+  // an acint to count from 0 to 'bits', inclusive
```
```suggestion
  // an ac_int to count from 0 to 'bits', inclusive

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_stacker.hpp:
```diff
@@ -0,0 +1,95 @@
+template <typename InPipe, typename OutPipe, unsigned literals_per_cycle>
+void ByteStacker() {
+  using OutPipeBundleT = decltype(OutPipe::read());
+  constexpr int cache_idx_bits = fpga_tools::Log2(literals_per_cycle * 2) + 1;
```
kCacheIdxBits

tyoungsc replied:
This goes against what I was told by <person who does code formatting stuff>. Since it is 'computed' (albeit at compile time), it should not be CamelCase. Those seem to be reserved for "Magic Numbers", like `constexpr int kMyMagicNum = 7;` I can use the `k_` prefix though. Does that work?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_stacker.hpp:
```diff
@@ -0,0 +1,95 @@
+    // try to read in some data
+    bool data_valid;
+    auto pipe_data = InPipe::read(data_valid);
+    done = pipe_data.flag && data_valid;
```
'flag' seems like a poor choice of name.  What is this member variable 'flagging'?  Why not call it 'done' or 'last' or something descriptive?

tyoungsc replied:
It is meant to be generic. It is "data + a flag". Could be `done`, `valid`, etc.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_stacker.hpp:
```diff
@@ -0,0 +1,95 @@
+
+    // add the valid data we read in to the cache
+    if (data_valid && !done) {
+#pragma unroll
```
Just a style question - why do we tend to put these #pragma statements starting at column 1?  I think I'd rather see them indented to match the loop they are being applied to.

tyoungsc replied:
clang-format does this. Looks like a future release of clang is addressing this, but nothing to fix it yet. I tried for a while.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_stacker.hpp:
```diff
@@ -0,0 +1,95 @@
+      // decrement cache_idx by number of elements we read
+      // it is safe to always subtract literals_per_cycle since that will only
+      // happen on the last iteration of the outer while loop (when 'done'
+      // is true)
```
```suggestion
      // it is safe to always subtract literals_per_cycle since that can only
      // result in a negative number on the last iteration of the outer while loop (when 'done'
      // is true), at which point the value will never be used

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/byte_stacker.hpp:
```diff
@@ -0,0 +1,95 @@
+  } while (!done);
+
+  // notify downstream kernel that we are done
+  OutPipe::write(OutPipeBundleT(true));
```
This seems odd.  I'm assuming the type used for OutPipe must define a constructor that takes only a bool?

This means you have two write sites for the pipe.  Would it be better to change the protocol so that the 'done' flag is set on the last valid piece of data?  Then you would only need a single write site for your pipes, you just have to set 'done' to true/false as appropriate.

tyoungsc replied:
I agree, this would be nicer, but I will treat it as a future optimization.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/common.hpp:
```diff
@@ -0,0 +1,140 @@
+#include <CL/sycl/INTEL/ac_types/ac_int.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+#include "constexpr_math.hpp"
```
file location

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/common.hpp:
```diff
@@ -0,0 +1,140 @@
+  static_assert(max_length > 0);
+  static_assert(max_distance > 0);
+
+  static constexpr unsigned max_symbols = n;
```
Should these constexpr member variables use the kVariableName naming convention, or does it not apply to class/struct members?

tyoungsc replied:
They could use the `k_` prefix, but not CamelCase, as commented before.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/common.hpp:
```diff
@@ -0,0 +1,140 @@
+  LZ77InputData() {}
+
+  // indicates whether this is a symbol or {length, distance} pair
+  bool is_copy;
```
Based on the name (is_copy?), I don't understand if 'true' means it's a symbol or a length/distance pair.  Why not 'is_symbol' or 'is_pair'?  Even if there's a good explaniation, please update the comment to indicate which is which.

tyoungsc replied:
will change to `is_literal` (and reverse the logic everywhere).

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/common.hpp:
```diff
@@ -0,0 +1,140 @@
+
+//
+// Holds an array of bytes, where valid_count indicates how many of the 'n'
+// bytes are valid. The valid bytes must be sequential.
```
```suggestion
// Holds an array of bytes, where valid_count indicates how many of the 'n'
// bytes are valid. The valid bytes must be sequential and start at index 0.

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_decompressor.hpp:
```diff
@@ -0,0 +1,85 @@
+
+#include "byte_stacker.hpp"
+#include "common.hpp"
+#include "constexpr_math.hpp"
```
file location comment

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_decompressor.hpp:
```diff
@@ -0,0 +1,85 @@
+using namespace sycl;
+
+// declare the kernel and pipe names globally to reduce name mangling
+class GzipMetadataReaderKernelID;
+class HuffmanDecoderKernelID;
+class LZ77DecoderKernelID;
+class ByteStackerKernelID;
+
+class GzipMetadataToHuffmanPipeID;
+class HuffmanToLZ77PipeID;
+class LZ77ToByteStackerPipeID;
```
Is it worth templating these class names on an integer to allow a design to easily instantiate multiple gzip_decompressors?

tyoungsc replied:
Yes. Will do.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_decompressor.hpp:
```diff
@@ -0,0 +1,85 @@
+                           HuffmanToLZ77Pipe>(q);
+
+  // the design only needs a ByteStacker kernel when literals_per_cycle > 1
+  if constexpr (literals_per_cycle > 1) {
```
Have you tested the code with literals_per_cycle = 1?

tyoungsc replied:
Yes.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_header_data.hpp:
```diff
@@ -0,0 +1,154 @@
+//
+// States for parsing the GZIP header
+//
+enum GzipHeaderState {
```
Why not use an enum class here?  https://www.geeksforgeeks.org/enum-classes-in-c-and-their-advantage-over-enum-datatype/

tyoungsc replied:
Probably not necessary in this case, but done.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_header_data.hpp:
```diff
@@ -0,0 +1,154 @@
+  std::string Filename() const {
+    std::string ret;
+    int i = 0;
+    while (i < 256 && filename[i] != '\0') {
```
Declare kMaxFilenameLength = 256 and use that instead of a 'magic number' here.

Why are we limiting filename to 256 characters?  Is that part of the GZIP spec?

tyoungsc replied:
Will do.

> Why are we limiting filename to 256 characters? Is that part of the GZIP spec?

Yes.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_header_data.hpp:
```diff
@@ -0,0 +1,154 @@
+      case 0:
+        return "FAT";
+      case 1:
+        return "Amiga";
```
Wow, this is old

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/gzip_header_data.hpp:
```diff
@@ -0,0 +1,154 @@
+      case 4:
+        return "VM/CMS";
+      case 5:
+        return "Atari TOS";
```
Atari?!?

tyoungsc replied:
hahaha

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+
+// maximum bits to read dynamically from the bitstream
+// more can be read statically using the templated Shift function
+constexpr int kBitBufferMaxReadBits = 5;
```
Add a comment explaining why 5 is the correct number here, based on the DEFLATE algorithm?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+constexpr int kBitBufferMaxReadBits = 5;
+
+// maximum bits consumed in one shift of the bitstream
+constexpr int kBitBufferMaxShiftBits = 30;
```
Add a comment explaining the number 30?

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+  bool last_block;
+  bool done_reading = false;
+
+  // processing consecutive blocks
```
I think this comment is meant to explain why we disable loop pipelining, but I don't understand the comment

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+      dist_map_first_code[codelen - 1] = dist_map_next_code;
+      dist_map_base_idx[codelen - 1] = dist_map_counter;
+
+      // this loop has been manually fused for the literal and distance codes
```
Why do manual fusing instead of writing the loops intuitively and applying loop fusion in the compiler?

tyoungsc replied:
I like the manual fusing because it is more intuitive to me.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+    // indicates whether we are reading a distance (or literal) currently
+    bool reading_distance = false;
+
+    // true is the stop code (256) has been decoded and the block is done
```
```suggestion
    // true if the stop code (256) has been decoded and the block is done

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+    // the -DHUFFMAN_II=<desired II>. By default, we let the
+    // the compiler choose the Fmax/II to maximize throughput
+#ifdef HUFFMAN_MAIN_LOOP_II
+    [[intel::initiation_interval(HUFFMAN_II)]]  // NO-FORMAT: Attribute
```
Why leave it up to the compiler by default?  Doesn't this expose us to major throughput degredation from some future compiler change?

tyoungsc replied:
This loop is throughput bottleneck and,  in a perfect world, the compiler would choose the best Fmax/II combination for throughput.

tyoungsc replied:
Which, for all my tests to HW, it has been doing.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+        if (!reading_distance) {
+          shift_amount = lit_shortest_match_len;
+          // currently parsing a symbol or length (same table)
+          if (lit_symbol == 256) {
```
There are a lot of 'magic numbers' in here.  Wouldn't it be better to define named constants?

tyoungsc replied:
I don't think named constants are always better. In this case, I think it would just be a distraction.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/huffman_decoder.hpp:
```diff
@@ -0,0 +1,750 @@
+  ////////////////////////////////////////////////////////////////////////////
+
+  // notify the downstream kernel that we are done
+  OutPipe::write(OutPipeBundleT(true));
```
I made a comment elsewhere about changing the protocol to make the flag indicate 'last' instead of 'done', which allows a single pipe write site to save area.  Same comment applies here.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,436 @@
+// which contains either a literal from upstream, or a {length, distance} pair.
+// Given a literal input, this function simply tracks that literal in a history
+// buffer and writes it to the output. For a {length, distance} pair, this
+// this function reads 'literals_per_cycle' elements from the history buffer
```
```suggestion
// function reads 'literals_per_cycle' elements from the history buffer

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,436 @@
+
+  // we will cyclically partition the history to 'literals_per_cycle' buffers,
+  // so each buffer gets this many elements
+  constexpr size_t history_buffer_count = max_history / literals_per_cycle;
```
Lots of constexpr variables here that don't use the kVariableName naming convention

tyoungsc replied:
Same response as above. This goes against what I was told by formatting folk.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,436 @@
+  HistBufBufIdxT read_history_shuffle_idx[literals_per_cycle];
+
+  // precompute the function: dist + ((i - dist) % dist)
+  // which is used for the special when the copy distance is less than
```
```suggestion
  // which is used for the corner case when the copy distance is less than

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/lz77_decoder.hpp:
```diff
@@ -0,0 +1,436 @@
+}
+
+//
+// Top-level LZ77 decoder to switch between the single- and multi-element per
```
```suggestion
// Top-level LZ77 decoder to select between the single- and multi-element per

```
'switch' implies to me that it's happening dynamically at runtime

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/simple_crc32.hpp:
```diff
@@ -0,0 +1,36 @@
+#define __SIMPLE_CRC32_HPP__
+
+//
+// A simple, non-performant CRC-32 implementation.
```
```suggestion
// A simple CRC-32 implementation (not optimized for high performance).

```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/main.cpp:
```diff
@@ -0,0 +1,500 @@
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities/include/dpc_common.hpp
+#include "common.hpp"
+#include "constexpr_math.hpp"
```
file location comment

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/main.cpp:
```diff
@@ -0,0 +1,500 @@
+  // the device selector
+#if defined(FPGA_EMULATOR)
+  sycl::ext::intel::fpga_emulator_selector selector;
+#elif defined(FPGA_SIMULATOR)
```
Something to think about for the future: should we add sim target support to dpc_common.hpp?  Or if not, create a shared header file in the FPGA include directory with the required supporting classes/functions?

tyoungsc replied:
Agreed.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/src/main.cpp:
```diff
@@ -0,0 +1,500 @@
+      std::cout << "All kernels have finished for run " << i << std::endl;
+
+      // duration in milliseconds
+      time_ms[i] = duration<double, std::milli>(end - start).count();
```
Aren't there SYCL runtime functions that we're supposed to use to 'properly' measure kernel execution times?

tyoungsc replied:
Yes. Don't like them, and std::chrono is just as accurate ...  our SYCL timing is ... not great.

KanclerzPiotr committed:
ONSAM-1467 atomic_ref is no longer extension (#901)

atomic_ref is no longer extension

louie-tsai committed:
Update requirements.txt (#904)

add runipy for ONSAM-1465

praveenkk123 committed:
updated Welcome.ipynb and Readme to address ONSAM-1451 to include ste… (#899)

* updated Welcome.ipynb and Readme to address ONSAM-1451 to include steps to run the notebooks locally and ONSAM-1450 to fix broken links in welcome.ipynb

* Update README.md

tyoungsc committed:
First commit of unified GZIP and SNAPPY designs into one reference design

tyoungsc committed:
Addressed some of Mike's comments on the old gunzip in the new files

tyoungsc committed:
Removing gunzip directory in favour of new decompress unified design

tyoungsc committed:
Switch to only while-loops

tyoungsc committed:
Big code cleanup. Moved parsing of first and second Huffman tble to separate functions in Huffman decoder

yuguen commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/README.md:
````diff
@@ -0,0 +1,427 @@
+     ```
+
+### Example of Output
+You should see output similar to the following in the console for the GZIP and SNAPPY versions, respectively:
````
How is the user selecting between a GZIP and SNAPPY decompression? Is it chosen at compile time? or are both versions compiled to HW? Does the exe figures out if it needs to decompress using GZIP or SNAPPY?

tyoungsc replied:
Using a compile-time macro. From main.cpp:
```cpp
// ensure only one of GZIP and SNAPPY is defined
#if defined(GZIP) and defined(SNAPPY)
static_assert(false, "Only one of GZIP and SNAPPY can be defined!");
#endif

// if neither of GZIP and SNAPPY is defined, default to SNAPPY
#if not defined(GZIP) and not defined(SNAPPY)
#define SNAPPY
#endif
```

I will add a not about this in the README

tyoungsc replied:
I'll also add a line in the stdout saying which version of the decompression is happening.

mdbtucker replied:
I think you should probably have a CMAKE argument that adds the appropriate command line arguments.  That is what I did for gzip/gzip_ll.  You could make it just a makefile argument, but it fits better with our regtest infra if it's done in the cmake (not the best reason to do it that way, but I like the cmake solution anyway)

tyoungsc replied:
I do have this. But I included this in case someone goes rogue and doesn't use CMake.

louie-tsai committed:
fix for ONSAM-1476  (#908)

* fix an compilation error

* a fix for compiling errror for 2022.2

* error handling

Commit:
Clean metadata on SKLearnEx GS sample causing errors (#906)

* try fixing GS SKlearnex sample

* clean sklearnex metadata

* improve testing

Co-authored-by: u60279 <u60279@s001-n001.aidevcloud>

louie-tsai committed:
changes for oneccl 2021.6 release (#870)

* Update README.md

* Update README.md

louie-tsai committed:
fix for ONSAM-1442 (#912)

* remove the horovd sample

* update README for horovod sample removal

tyoungsc committed:
Addressed Yohann's comments about selecting between GZIP and SNAPPY
Added stdout message to say whether GZIP or SNAPPY is running
Added to README to indicate how to run cmake for GZIP vs SNAPPY
Added metaprogramming util to check if a type is a SYCL pipe (with any templating)
Added check to all kernels to ensure InPipe and OutPipe are SYCL pipes
Cleaned up static_asserts in metaprogramming_utils
clang-format on metapgramming_utils.hpp

tyoungsc committed:
Improved check in memory_utils.hpp to ensure pipe and pointer type have the same base value

KanclerzPiotr committed:
Fix for ONSAM-1480 (#914)

Change linker optimization for release mode

louie-tsai committed:
remove windows support for ONSAM-1471 and merge samples into one  (#913)

* remove windows support and add all cpp files from installation

* merge two oneDNN samples into GSG

* remove windows support

* Update README.md to address the changes

* Update README.md

* switch to opencl backend

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/README.md:
```diff
@@ -1,5 +1,5 @@
-# GZIP Decompression
-This DPC++ reference design implements GZIP decompression on an FPGA. This includes LZ77 decoding and Huffman decoding with static or dynamic huffman tables. 
+# Decompression
+This DPC++ reference design implements both GZIP and Snappy decompression on an FPGA.
```
```suggestion
This DPC++ reference design can be compiled to implement either GZIP or Snappy decompression on an FPGA.
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/README.md:
```diff
@@ -101,7 +101,7 @@ After learning how to use the extensions for Intel oneAPI Toolkits, return to th
        make fpga
```
The instructions should be updated to tell you how to choose between GZIP and SNAPPY.  I'm assuming (hoping) it's an argument to CMAKE . . .

tyoungsc replied:
It is an argument to CMake, and the instructions are in the README:


```
To select between GZIP and SNAPPY decompression, use the -DGZIP=1 and -DSNAPPY=1 `cmake` flags as shown below, respectively:
 
   cmake .. -DGZIP=1
   cmake .. -DSNAPPY=1
```

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/CMakeLists.txt:
```diff
@@ -14,6 +14,25 @@ else()
 endif()
 
+# Select between SNAPPY and GZIP decompression
+if(DEFINED SNAPPY AND DEFINED GZIP)
```
I think by checking if these are defined, you will generate some CMAKE warnings about not using the values.  I'm pretty sure an undefined variable evaluates to FALSE, so you can just do if (SNAPPY AND GZIP).  This would also allow you to do cmake -DSNAPPY=0 -DGZIP=1 and have that work (not that this is terribly helpful...)

tyoungsc replied:
Addressed below.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/CMakeLists.txt:
```diff
@@ -14,6 +14,25 @@ else()
+    set(DECOMPRESS_FORMAT_FLAG "-DSNAPPY")
+  endif()
+
+  set(ignoreMe "${GZIP}${SNAPPY}")
```
oooh, I haven't seen this syntax before - does this get rid of the warnings?  The solution I propose above works too, but as long as this eliminates the CMAKE warnings I'm fine with it.

tyoungsc replied:
yes. It is intended to get rid of warnings. And it works.

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/common/byte_stacker.hpp:
```diff
@@ -1,13 +1,14 @@
 #ifndef __BYTE_STACKER_HPP__
 #define __BYTE_STACKER_HPP__
 
+// clang-format off
```
What's the justification for this?

tyoungsc replied:
I want to pick the order of my includes to be batched together and make more sense. For example, this formatting gets really messed up by clang-format:

```cpp
#include <CL/sycl.hpp>
#include <sycl/ext/intel/ac_types/ac_int.hpp>
#include <sycl/ext/intel/fpga_extensions.hpp>

// Included from DirectProgramming/DPC++FPGA/include/
#include "constexpr_math.hpp"
#include "metaprogramming_utils.hpp"
```

Especially you start using std:: headers too.

mdbtucker replied:
Why not just add the commend on the same line as each header it applies to?  For example:

#include "common.hpp"
#include "constexpr_math.hpp"   // Included from DirectProgramming/DPC++FPGA/include/
#include "more_local_stuff.hpp"

But your solution is also fine, just a suggestion.

tyoungsc replied:
This line you just sent:

```cpp
#include "constexpr_math.hpp" // Included from DirectProgramming/DPC++FPGA/include/
```

Is 83 characters. That is why.

mdbtucker replied:
#include "constexpr_math.hpp" // from DirectProgramming/DPC++FPGA/include/

mdbtucker replied:
Or:

#include "constexpr_math.hpp" // from ../../include

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/snappy/snappy_data_gen.hpp:
```diff
@@ -0,0 +1,159 @@
+//    'num_copies' of length 'copy_len' and offset max(16k, lit_str_len)
+//    'repeat' copies of the above.
+//
+std::vector<unsigned char> GenerateSnappyCompressedData(unsigned lit_str_len,
```
Why do you need this for Snappy but not for GZIP?

Do you have an independent source of Snappy compressed data, to make sure you're complying with the standard correctly?  By writing your own test data, you risk covering up some misunderstanding of the format.

tyoungsc replied:
I can include one more snappy test that uses a real file.

tyoungsc committed:
Addressing some of Mike's comments:
- Adding test to decompress a Snappy file that is not generated from my program
- Changing README to account for this test

Other things:
- Added a ton more static_assert checks to validate input and output data of blocks

yuguen committed:
FPGA QRD update seeds and improve program duration computation (#917)

* use the events to time the execution

Signed-off-by: Yohann Uguen <yohann.uguen@intel.com>

* Update seeds

Signed-off-by: Yohann Uguen <yohann.uguen@intel.com>

tyoungsc committed:
Added link to on-chip caching tutorial in README

tyoungsc committed:
Added a bunch of comments above many functions/classes to explain template parameters and arguments
Added more static asserts

karlqi2000 committed:
Removing Function Pointer from CI (#918)

* Regression in how compiler handles function pointers on GPU. Removing
prog2 from CI automated test

* Revert "Regression in how compiler handles function pointers on GPU. Removing"

This reverts commit 9e09fc89a0bdb99fff7b7a04c9b645f2185578a2.

* Due to Compiler Bug, ensure to run function pointer only on CPU

tyoungsc committed:
Added comments
Comment grammar
Code formatting and cleanup, variable names update

tyoungsc committed:
Line in README about include directory files
Updated samples.json for include directory files

tyoungsc committed:
Switched enum to enum class
Made hyperlink text in README more informative
